feat: Export as Delta/PDF/PlainText functionality.

mmjsmohit opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This PR aims to add the export functionality to this app. This PR contains three options in the export menu - Export as Delta Object, Export as Plain Text and Export as a PDF.
## Featured Covered in this PR
- [x] Delta Object
- [X] Plain Text
- [X] PDF

## Related Tickets & Documents

<!--
For pull requests that relate or close an issue, please include them
below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Closes #25 

##Screenshots, Recordings 
![WhatsApp Image 2023-10-02 at 11 54 59 AM](https://github.com/SankethBK/diaryvault/assets/65406090/4c2cf6e5-ce67-4355-bf65-40c1fb23cf8a)
![image](https://github.com/SankethBK/diaryvault/assets/65406090/2635af60-389d-4426-83d3-6a261a59634f)

## Tested Feature??

- [X] In Real Device.
- [X] In Emulator

SankethBK commented:
Hey getting a blank popup here

[export_to_pdf.webm](https://github.com/SankethBK/diaryvault/assets/51091231/de1a7801-1336-4d4e-b9d7-ce875b445dce)

mmjsmohit commented:
I don't think its a blank popup, I think the text and icon color are also black which I will fix in a commit shortly.

mmjsmohit commented:
Hi! I have added dynamic styling to the share sheet. Now the text is properly visible as it is pulling the colour directly from the current theme.

The merge conflict in the `dropbox_sync_client.dart` is likely due to fact that I have fixed a typo.

micedreams commented on lib/features/notes/core/converters/delta_to_pdf.dart:
```diff
@@ -0,0 +1,146 @@
+        }
+      });
+    }
+    return [fontWeight, fontSize];
```
please add a white line before the return statement.

micedreams replied:
```suggestion

    return [fontWeight, fontSize];
```
Please add white line before the return statement.

micedreams commented on lib/features/notes/core/converters/delta_to_pdf.dart:
```diff
@@ -0,0 +1,146 @@
+            break;
+        }
+      });
+    }
```
[same here](https://github.com/SankethBK/diaryvault/pull/45/files#r1349385722)

micedreams commented on lib/features/notes/core/converters/delta_to_pdf.dart:
```diff
@@ -0,0 +1,146 @@
+      fontWeight = pw.FontWeight.bold;
+      fontSize = 16;
+    }
+    return [fontWeight, fontSize];
```
Please add white line before the return statement.
```suggestion

    return [fontWeight, fontSize];
```

micedreams commented on lib/features/notes/core/converters/delta_to_pdf.dart:
```diff
@@ -0,0 +1,146 @@
+  /// [Returns] fontWeight & fontSize
+  getPDFHeaderStyle(key, value) {
+    pw.FontWeight fontWeight = pw.FontWeight.normal;
+    double fontSize = 14;
```
```suggestion
    double fontSize = 14;
    
```
Please add a white line after all declaration statements.

micedreams commented on lib/features/notes/core/converters/delta_to_pdf.dart:
```diff
@@ -0,0 +1,146 @@
+  getHeaderAttributedText(
+      Map<String, dynamic>? attribute, String text, bool hasAttribute) {
+    pw.FontWeight fontWeight = pw.FontWeight.normal;
+    double fontSize = 14;
```
```suggestion
    double fontSize = 14;
    
```

micedreams commented on lib/features/notes/core/converters/delta_to_pdf.dart:
```diff
@@ -0,0 +1,146 @@
+        }
+      });
+    }
+    return pw.Text(
```
```suggestion

    return pw.Text(
```

micedreams commented on lib/features/notes/core/converters/delta_to_pdf.dart:
```diff
@@ -0,0 +1,146 @@
+  deltaToPDF(List<fq.Operation> deltaList) {
+    List header = [null, null];
+    List texts = [];
+    List pdfColumnWidget = [];
```
```suggestion
    List pdfColumnWidget = [];
    
```

micedreams commented on lib/features/notes/core/converters/delta_to_pdf.dart:
```diff
@@ -0,0 +1,146 @@
+    PdfColor fontColor = PdfColor.fromHex("#000");
+    pw.FontStyle fontStyle = pw.FontStyle.normal;
+    pw.TextDecoration decoration = pw.TextDecoration.none;
+    pw.BoxDecoration boxDecoration = const pw.BoxDecoration();
```
```suggestion
    pw.BoxDecoration boxDecoration = const pw.BoxDecoration();
    
```

micedreams commented on lib/features/notes/core/converters/delta_to_pdf.dart:
```diff
@@ -0,0 +1,146 @@
+        [
+          pw.Column(children: [...texts.reversed])
+        ];
+    return pw.Padding(
```
```suggestion

    return pw.Padding(
```

micedreams commented on lib/features/notes/presentation/widgets/note_export_button.dart:
```diff
@@ -0,0 +1,122 @@
+            padding: const EdgeInsets.only(right: 13.0),
+            child: IconButton(
+              icon: const Icon(Icons.share),
+              onPressed: () {
+                _showBottomSheet(context, state);
+              },
```
```suggestion
              onPressed: () => _showBottomSheet(context, state),
```

micedreams commented on lib/features/notes/presentation/widgets/note_export_button.dart:
```diff
@@ -0,0 +1,122 @@
+    builder: (BuildContext context) {
+      final mainTextColor =
+          Theme.of(context).extension<PopupThemeExtensions>()!.mainTextColor;
+      return Container(
```
```suggestion

      return Container(
```

micedreams commented on packages/flutter_quill/flutter_quill_extensions/lib/embeds/builders.dart:
```diff
@@ -73,72 +73,72 @@ class ImageEmbedBuilder extends EmbedBuilder {
+                  icon: Icons.settings_outlined,
+                  color: Colors.lightBlueAccent,
+                  text: 'Resize'.i18n,
+                  onPressed: () {
+                    Navigator.pop(context);
+                    showCupertinoModalPopup<void>(
+                        context: context,
+                        builder: (context) {
+                          final _screenSize = MediaQuery.of(context).size;
+                          return ImageResizer(
+                              onImageResize: (w, h) {
+                                final res = getEmbedNode(
+                                    controller, controller.selection.start);
+                                final attr = base.replaceStyleString(
+                                    getImageStyleString(controller), w, h);
+                                controller
+                                  ..skipRequestKeyboard = true
+                                  ..formatText(
+                                      res.offset, 1, StyleAttribute(attr));
+                              },
+                              imageWidth: _imageSize?.width,
+                              imageHeight: _imageSize?.height,
+                              maxWidth: _screenSize.width,
+                              maxHeight: _screenSize.height);
+                        });
```
Please turn this into a separate method.
Nesting too many times is not a good idea...

https://www.youtube.com/watch?v=CFRhGnuXG-4

micedreams commented on packages/flutter_quill/flutter_quill_extensions/lib/embeds/builders.dart:
```diff
@@ -73,72 +73,72 @@ class ImageEmbedBuilder extends EmbedBuilder {
+                    Navigator.pop(context);
+                  },
+                );
+                return Padding(
```
```suggestion

                return Padding(
```

micedreams reviewed (commented):
Minor changes and this PR is probably good to go!!🎉

mmjsmohit commented:
Thanks for the formatting tips. I think you can resolve the merge conflicts within the `dropbox_sync_client.dart` @SankethBK

SankethBK commented:
@mmjsmohit slight change of plans, we added text to speech button in place of the share button that you currently have.
<img src="https://user-images.githubusercontent.com/51091231/272696858-130e9b11-d299-4f41-89b5-3bb9b7859a06.jpeg" width = "500" />. 
I think the best place to have the share button here would be next to delete button in home page. So that users can export multiple notes as PDF at once. 

The main logic for PDF generation will still remain same as you implemented. It's just that we'll be moving `NoteShareButton` to home page app bar next to delete button. There will also be slight code changes related to generating the PDF for multiple notes instead of one. 

I will merge your PR to `export_as_pdf` branch which can be used as starting point for next changes, i will raise separate issue for this or refactor it myself. 

Thanks a lot for your contribution, I think this is the toughest functionality implemented in this app so far.
